docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,119 @@
+Once they have a complete state, they can use "Block Sync" to download
+and execute any more recent blocks.
+
+Note that Tendermint does not actually know anything about the application
+state; it is only available over ABCI. State Sync will thus require new ABCI
+methods and/or connections, and will need to support arbitrary Merkle trees.
+
+### Prior Work
+
+There has been some related work in industry, though with different security
+model and architecture. TODO.
+
+Parity Ethereum implements a ["Warp
+Sync"](https://wiki.parity.io/Warp-Sync-Snapshot-Format.html) to rapidly
+download both blocks and state snapshots from peers. Data is carved into ~4MB
+chunks and snappy compressed. Hashes of snappy compressed chunks are stored in a
+manifest file which co-ordinates the state-sync. Obtaining a correct manifest
+file seems to require an honest majority of peers. This means you may not find
+out the state is incorrect until you download the whole thing and compare it
+with a verified block header.
+
+A similar solution was implemented by Binance in
+[#3594](https://github.com/tendermint/tendermint/pull/3594)
+based on their initial implementation in
+[PR #3243](https://github.com/tendermint/tendermint/pull/3243)
+and [some learnings](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit).
+Note this still requires the honest majority peer assumption.
+One major advantage of the warp-sync approach is that chunks can be compressed before sending.
+
+Ideally State Sync in Tendermint would not require an honest majority of peers, and instead
+would rely on the standard light client security, meaning state chunks would be
+verifiable against the merkle root of the state. Such a design for tendermint
+was originally tracked in [#828](https://github.com/tendermint/tendermint/issues/828).
+
+An [initial specification](https://docs.google.com/document/d/15MFsQtNA0MGBv7F096FFWRDzQ1vR6_dics5Y49vF8JU/edit?ts=5a0f3629) was published by Alexis Sellier.
+In this design, the state has a given `size` of primitive elements (like keys or
+nodes), each element is assigned a number from 0 to `size-1`,
+and chunks consists of a range of such elements.
+Ackratos raised
+[some concerns](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit)
+about this design, somewhat specific to the IAVL tree, and mainly concerning
+performance of random reads and of iterating through the tree to determine element numbers
+(ie. elements aren't indexed by the element number).
+
+This design could be generalized to have numbered chunks instead of elements,
+where chunks are indexed by their chunk number and determined by some algorithm
+for grouping the elements in the tree. This would allow pre-computation and
+persistence of the chunks that could alleviate the concerns about real-time iteration and random reads.
+However, there seems to be a larger issue with the "numbered chunks" approach:
```

Comment:
After further thought it seems definitely possible to do this, but it might require that the first few chunks come in order and provide information about how the rest of the chunks are indexed. For instance the first chunk might provide the first 5 layers of the tree, and then there might be 32 more chunks, one for each subtree. Those would then be verifiable against their index without any more commitments.

Reply:
cc https://github.com/tendermint/iavl/issues/135